modules/contracts/apps/challenge/src/handlers/execute.rs

```diff
@@ -0,0 +1,568 @@
+) -> AppResult {
+    // Ensure the caller is an admin
+    app.admin.assert_admin(deps.as_ref(), &info.sender)?;
+
+    let mut existing_friends = CHALLENGE_FRIENDS
+        .may_load(deps.storage, challenge_id.clone())?
+        .unwrap_or_default();
+
+    for friend in &friends {
+        if existing_friends.iter().any(|f| f.address == friend.address) {
+            return Err(AppError::AlreadyAdded {});
+        }
+    }
+
+    // validate the String addresses and convert them to Addr
+    // before saving
+    let friends: Vec<Friend<Addr>> = friends
+        .iter()
+        .cloned()
+        .map(|friend| friend.check(deps.as_ref()).unwrap())
+        .collect();
+
+    existing_friends.extend(friends);
+    CHALLENGE_FRIENDS.save(deps.storage, challenge_id, &existing_friends)?;
+    Ok(Response::new().add_attribute("action", "add_friends"))
+}
+
+pub fn remove_friends_from_challenge(
+    deps: DepsMut,
+    info: MessageInfo,
+    app: &ChallengeApp,
+    challenge_id: u64,
+    friend_addresses: Vec<Friend<String>>,
+) -> AppResult {
+    // Ensure the caller is an admin
+    app.admin.assert_admin(deps.as_ref(), &info.sender)?;
+
+    let mut existing_friends = CHALLENGE_FRIENDS
+        .may_load(deps.storage, challenge_id.clone())?
+        .unwrap_or_default();
+
+    for friend in &friend_addresses {
+        if !existing_friends.iter().any(|f| f.address == friend.address) {
+            return Err(AppError::Std(StdError::generic_err(
+                "Friend not found for this challenge",
+            )));
+        }
+    }
+
+    existing_friends.retain(|f| {
+        !friend_addresses
+            .iter()
+            .any(|friend| f.address == friend.address)
+    });
+    CHALLENGE_FRIENDS.save(deps.storage, challenge_id, &existing_friends)?;
+    Ok(Response::new().add_attribute("action", "remove_friends"))
+}
+
+fn daily_check_in(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    app: &ChallengeApp,
+    challenge_id: u64,
+    metadata: Option<String>,
+) -> AppResult {
+    app.admin.assert_admin(deps.as_ref(), &info.sender)?;
+    let mut challenge = CHALLENGE_LIST.load(deps.storage, challenge_id.clone())?;
+    let now = Timestamp::from(env.block.time);
+
+    // If the challenge has ended, we set the status to OverAndPending and return before checking
+    // in.
+    if now > challenge.get_end_timestamp()? {
+        match challenge.status {
+            ChallengeStatus::Active => {
+                challenge.status = ChallengeStatus::OverAndPending;
+                CHALLENGE_LIST.save(deps.storage, challenge_id.clone(), &challenge)?;
+                return Ok(Response::new()
+                    .add_attribute("action", "check_in")
+                    .add_attribute(
+                        "message",
+                        "Challenge has ended. You can no longer register a daily check in. 
+                         ChallengeStatus has been set to OverAndPending.",
+                    ));
```

Comment:
We should use the app.tag_response here so that the wasm-abstract event is included.

Reply:
Same comment below

Reply:
We can also create the response at the top of the method, mutating it throughout, then return it at the very bottom.